readme.md
Hue Control
===========

> Note that I am not actually using (and therefore maintaining) this addon anymore.

About
-----
Hue Control is a Kodi addon that can change your hue lights when starting a movie.
Next to that it has a few preset slots that you can use to recall any lightsetting you created earlier with your TV remote.


Installation
------------
Just install from the zip file
When updating the addon make sure to disable and enable the addon listed under Services/huecontrol because Kodi does not restart it automatically (Restarting Kodi also works).

If you have problems upgrading the addon then:
- Uninstall the addon
- Restart Kodi
- Install the addon


Setup
-----
In the addons settings screen you will fist have to connect to your bridge. Follow the instructions from the addon to go through the procedure.

You can select which lamps to control in the settings window (e.g. only the lamps in the livingroom). By default 3 lamps are selected.

By default the following behaviours are configured:
- When playing a movie the lights will go out
- When pausing a movie the lights will go to a low brightness setting
- When stopping a movie the lights return to their original state.

If you want to change the settings of the lamps go to the "Presets" section in the settings.

1. Locate the preset you want to change. 
The "Playing" and "Paused" presets will be automatically selected when playing/pausing a movie. The other presets are available in the presets menu that can be opened when opening the addon from te programs section of Kodi or you can bind it to a key on your remote control through the keymap.xml by adding the following for a key:
> RunScript(service.huecontrol,action=showpresets)

2. Set the lamps to the desired settings with an app (e.g. the Philips hue app)

3. Press the 'save lamp state' button for the preset under which you want to store the configuration. 




changelog.txt
0.10
 - Extend minimum-video-length range to 90 minutes

0.9
Note: Some settings will be lost when updating to this version. Please reconfigure the addon
- Change XBMC references to Kodi
- Minimum supported version of Kodi is now 14 (Helix)
- Add support for up to 63 lights
- Update addon icon and background
- Add debug logging (when Kodi debug logging is enabled)
- Misc. updates to make the addon more futureproof
- Use requests for network requests
- Improve searching for bridges
- Add support for migrated bridges
- Update bridge authentication code for upcoming changes in the Hue API
- Bugfixes

0.8
- Reworked settings handling (old settings will be lost)
- Reorganised settings window
- Added a readme

0.7
- Refactored bridge code
- Added presets, you can select them when starting te addon from the program sections or you can bind it
  to a button with the following code "RunScript(service.huecontrol,action=showpresets)"

0.6
- Wrong lamps were controlled (broken since 0.4)

0.5
- Last lamp was not controlled
- Extended number of lamps to 15
- Show settings when starting the addon from the programs section of XBMC

0.4
- Fixed lights not returning to saved state when movie ends
- Updated to the new XBMC translation system

0.3:
- Changed settings handling so now you can link a bridge and immediately save the lampstate (without leaving the settings window)
- More translatable strings
- Use own images

0.2:
- Update for Frodo

0.1:
- Initial version
 


huecontrol_test.py
#
# Code for testing stuff in the hue.py file
#

import hue
import time
import huecontrol


BRIDGEID = "0017880a88ea"
BRIDGEIP = "192.168.178.26"

# To skip the searching tests as they take long
skipSearching = True

if not skipSearching:
    bridge = hue.BridgeLocator().FindBridgeById(BRIDGEID)

    if not bridge:
        print "ERROR: Bridge not found"
    else:
        print "Found bridge: "
        print bridge

    
    bridge = hue.BridgeLocator("1.2.3.4").FindBridgeById(BRIDGEID)

    if not bridge:
        print "Bridge not found (as expected)"
    else:
        print "ERROR: Found bridge: "
        print bridge


    bridge = hue.BridgeLocator(BRIDGEIP).FindBridgeById(BRIDGEID)

    if not bridge:
        print "ERROR: Bridge not found"
    else:
        print "Found bridge: "
        print bridge


    bridges = hue.BridgeLocator().FindBridges()

    if not bridges:
        print "ERROR: No bridges found"
    else:
        print "Found bridges: "
        print bridges


bridge = hue.BridgeLocator(BRIDGEIP).FindBridgeById(BRIDGEID)

if not bridge:
    print "ERROR: Bridge not found"
else:
    print "Found bridge: "
    print bridge

    
def log(message):
    print message
    
bridge.username = "aValidUser"
bridge.devicetype = huecontrol.DEVICETYPE
bridge.logfunc = log


#for bridge in bridges:
print bridge.isAuthorized()

# Expect 101 error because the button is not pressed
result = bridge.authorize()
print result
if result != 101:
    print "Unexpected result " + str(result)[:100]


fullstate = bridge.getFullState()
print "First 250 chars of state:\n" + str(fullstate)[:250]

# Dim lights and restore light 1 and 3
bridge.PUT('/groups/0/action', {'bri':100})
bridge.setFullStateLights(fullstate, [1,3])

# Wait till restored
time.sleep(2)

# Now some without log function
bridge.logfunc = None


# Dim lights and restore all
bridge.PUT('/groups/0/action', {'bri':100})
bridge.setFullStateLights(fullstate)








script.py
#
# Kodi Script plugin that handles the actions for the huecontorl
#

import json
import httplib
import xbmc, xbmcgui, xbmcaddon
import subprocess,os
import sys
from urlparse import urlparse, parse_qs

import hue
import huecontrol
import xbmccommon
import time


def recallScene(sceneName):
    briOnly =  __addon__.getSetting("brightnessonly" + sceneName) == "true"
    state = hueAddonSettings.data["scene" + sceneName]
    logDebug("recall preset" + sceneName + ": " + str(state))

    bridge = hue.Bridge(ip=hueAddonSettings.data["bridgeip"], id=hueAddonSettings.data["bridgeid"],username=hueAddonSettings.data.get("bridgeusername", None))
    bridge.setFullStateLights(state,  xbmccommon.getConfiguredLampsList(), briOnly)


addonId = sys.argv[0]  # e.g.   service.huecontrol

__addon__ = xbmcaddon.Addon(id=xbmccommon.ADDON_ID)
__addonpath__ = xbmc.translatePath(__addon__.getAddonInfo('path')).decode("utf-8")  # Translate path to change special:// protocol to a normal path
__addonicon__ = os.path.join(__addonpath__, 'icon.png')
__language__ = __addon__.getLocalizedString


hueAddonSettings = xbmccommon.HueControlSettings()
logError = xbmccommon.logError
logDebug = xbmccommon.logDebug

idx = 1
parameters = {}
parameters["action"] = "none"

while idx < len(sys.argv):
    args = sys.argv[idx].split('=')
    parameters[args[0]] = args[1]
    idx += 1

if (parameters['action'] == "none"):
    # No action parameter, so must be run from programs thingy.
    # Lets show presets
    parameters['action'] = "showpresets"


    
if (parameters['action'] == "connect_to_bridge"):
    
    progress = xbmcgui.DialogProgress()
    progress.create(__language__(30007), __language__(30008))
    progress.update(0)
    
    bridges = hue.BridgeLocator(iprange=xbmc.getIPAddress(), logfunc=logDebug).FindBridges(progress=progress.update)
    bridgeidx = -1;
    
    progress.close();
    
    if (len(bridges) == 0):
        xbmcgui.Dialog().ok(__language__(30009), __language__(30010)) 
    elif (len(bridges) == 1):
        # Only one bridge, done
        bridgeidx = 0
        bridge = bridges[bridgeidx]
        xbmccommon.notify(__language__(30011).format(bridge.name)) # Keep output on one line. Name is name + IP e.g. Philips hue (111.112.113.114)
    else:
        dialog = xbmcgui.Dialog()
        
        bridgenames = ["{0} - {1} ({2})".format(bridge.name, bridge.id[-6:].upper(), bridge.ip) for bridge in bridges]
        bridgeidx = dialog.select(__language__(30020).format(len(bridgenames)), bridgenames)
    
    if (bridgeidx >= 0):

        bridge = bridges[bridgeidx]
        bridge.logfunc = logDebug
        if ("bridgeusername" in hueAddonSettings.data):
            bridge.username = hueAddonSettings.data["bridgeusername"]
        bridge.devicetype = huecontrol.DEVICETYPE.format(xbmc.getInfoLabel('System.FriendlyName') )
        
        xbmc.log(msg='Selected bridge {0} = {1}'.format(bridgeidx, bridge))
        
        hueAddonSettings.data["bridgeip"] = bridge.ip
        hueAddonSettings.data["bridgeid"] = bridge.id
        
        if (not bridge.isAuthorized()):
            # Perform authorization part
            # Use progress dialog to have a button with a cancel button
            progress = xbmcgui.DialogProgress()
            progress.create(__language__(30013), __language__(30014))
            progress.update(0)
            
            maxcount = 60
            count = 0
            while count < maxcount:
                progress.update(int((100.0/maxcount) * count), __language__(30014), __language__(30015).format(maxcount - count))
                result = bridge.authorize()
                
                if result == 0 or progress.iscanceled():
                    # done, break loop
                    count = maxcount
                
                count = count + 1
                time.sleep(1)
                
            progress.close();
            
        if (not bridge.isAuthorized()):
            xbmccommon.notify(__language__(30016), duration=5000)
        else:
            hueAddonSettings.data["bridgeusername"] = bridge.username
            # For safety remove any old (fixed) usernames
            bridge.DELETE("/config/whitelist/{0}".format(huecontrol.OLD_BRIDGEUSER))
            
            xbmccommon.notify(__language__(30017), duration=5000)
            
        hueAddonSettings.store()
    
elif (parameters['action'] == "savescene"):
    
    bridge = hue.Bridge(ip=hueAddonSettings.data["bridgeip"], id=hueAddonSettings.data["bridgeid"], username=hueAddonSettings.data.get("bridgeusername", None))
    
    state = bridge.getFullState()

    id = parameters['id']
    logDebug("save scene" + id + ": " + str(state))
    __addon__.setSetting("scene" + id, "") # I used to store the complete state here, write an empty string now, to "unbloat" the settings.xml
    hueAddonSettings.data["scene" + id] = state
    
    if hueAddonSettings.store():
        presetname = __addon__.getSetting("namescene" + id)
        if presetname == "":
            presetname = id
        xbmccommon.notify(__language__(30034).format(presetname))

elif (parameters['action'] == "recallscene"):

    recallScene(paramters['id'])

elif (parameters['action'] == "showpresets"):

    dialog = xbmcgui.Dialog()
    
    presetnames = []
    
    presetnames.append(__language__(30030))  # Playing
    presetnames.append(__language__(30040))  # Paused
    
    for i in range(huecontrol.NUM_PRESETS):
        logDebug("namescene" + str(i+1) + " - " +  __addon__.getSetting("namescenePreset" + str(i+1)))
        presetnames.append (__addon__.getSetting("namescenePreset" + str(i+1)))
    
    idx = dialog.select(__language__(30202), presetnames)
    
    if idx >= 0:
        # Assume one of the presets
        presetId = "Preset" + str(idx+1-2)
        
        if idx == 0:
            presetId = "Playing"
        if idx == 1:
            presetId = "Paused"

        recallScene(presetId)
    


service.py
#
# Kodi Service plugin that controls the brightness/on state of lamps
#

import json
import httplib
import xbmc,xbmcgui, xbmcaddon
import subprocess,os
import sys
from urlparse import urlparse, parse_qs

import hue
import huecontrol
import xbmccommon
import time

logDebug = xbmccommon.logDebug
logError = xbmccommon.logError
  
class HuePlayer(xbmc.Player):

    def __init__ (self):
        xbmc.Player.__init__(self)
        
        self.savedlampstate = {}
        self.CONTROLLING_LAMPS = 0
        self.addonId = xbmccommon.ADDON_ID

        logDebug("--> Init")

    def _setScene(self, scenename):
        __addon__ = xbmcaddon.Addon(id=self.addonId)
        hueAddonSettings = xbmccommon.HueControlSettings()
        
        self._setState(hueAddonSettings.data[scenename], __addon__.getSetting("brightnessonly" + scenename) == "true")

    def _setState(self, state, briOnly=False):
        __addon__ = xbmcaddon.Addon(id=self.addonId)

        hueAddonSettings = xbmccommon.HueControlSettings()
        bridge = hue.Bridge(ip=hueAddonSettings.data["bridgeip"], id=hueAddonSettings.data["bridgeid"], username=hueAddonSettings.data.get("bridgeusername", None))

        lamps = xbmccommon.getConfiguredLampsList()
        bridge.setFullStateLights(state, lamps, briOnly)

    def onPlayBackStarted(self):
        __addon__ = xbmcaddon.Addon(id=self.addonId)

        logDebug("--> onPlayBackStarted {0}, {1}".format(xbmc.Player().getTotalTime(), __addon__.getSetting("minvideolength")))

        if xbmc.Player().isPlayingVideo() and (xbmc.Player().getTotalTime() >= (float(__addon__.getSetting("minvideolength")) * 60) or xbmc.Player().getTotalTime() == 0):
            if (self.CONTROLLING_LAMPS == 0):
                hueAddonSettings = xbmccommon.HueControlSettings()

                bridge = hue.Bridge(ip=hueAddonSettings.data["bridgeip"], id=hueAddonSettings.data["bridgeid"], username=hueAddonSettings.data.get("bridgeusername", None))
                self.savedlampstate = bridge.getFullState()

            self.CONTROLLING_LAMPS = 1
            self._setScene("scenePlaying")
            
    
    def onPlayBackEnded(self):
        logDebug("--> onPlayBackEnded")
            
        if self.CONTROLLING_LAMPS == 1:
            self._setState(self.savedlampstate)
            
        self.CONTROLLING_LAMPS = 0

    def onPlayBackStopped(self):
        logDebug("--> onPlayBackStopped")

        if self.CONTROLLING_LAMPS == 1:
            self._setState(self.savedlampstate)

        self.CONTROLLING_LAMPS = 0
        
        
    def onPlayBackPaused(self):
        __addon__ = xbmcaddon.Addon(id=self.addonId)
        logDebug("--> onPlayBackPaused")

        if self.CONTROLLING_LAMPS == 1:
            self._setScene("scenePaused")


    def onPlayBackResumed(self):
        __addon__ = xbmcaddon.Addon(id=self.addonId)
        logDebug("--> onPlayBackResumed")

        if self.CONTROLLING_LAMPS == 1:
            self._setScene("scenePlaying")




# Check if the bridge still exists where we expect it
__addon__ = xbmcaddon.Addon(id=xbmccommon.ADDON_ID)
__language__ = __addon__.getLocalizedString


hueAddonSettings = xbmccommon.HueControlSettings()
hueBridgeOk = False

# Make sure the important settings exist
if ("bridgeip" in hueAddonSettings.data and "bridgeid" in hueAddonSettings.data):
    bridge = hue.BridgeLocator(iprange=xbmc.getIPAddress()).FindBridgeById(hueAddonSettings.data["bridgeid"], hueAddonSettings.data["bridgeip"])
    
    if bridge == None:
        xbmccommon.notify(__language__(30019), duration=10000)
    else:
        bridge.username = hueAddonSettings.data.get("bridgeusername", None)
        bridge.devicetype = huecontrol.DEVICETYPE.format(xbmc.getInfoLabel('System.FriendlyName'))
        
        if not bridge.isAuthorized():
            xbmccommon.notify(__language__(30019), duration=10000)
        else:
            hueAddonSettings.data["bridgeip"] = bridge.ip
            hueAddonSettings.data["bridgeid"] = bridge.id
            
            if hueAddonSettings.store():
                xbmccommon.notify(__language__(30018))
                

            
            
huePlayer = HuePlayer()
monitor = xbmc.Monitor()
 
while not monitor.abortRequested():
    # Sleep/wait for abort for 10 seconds
    if monitor.waitForAbort(10):
        # Abort was requested while waiting. We should exit
        break


xbmccommon.py
#
# Common code for Kodi related stuff
#

import xbmc, xbmcaddon
import pickle
import os
import errno
import hue
import huecontrol
import json

ADDON_ID = 'service.huecontrol'


__addon__ = xbmcaddon.Addon(id=ADDON_ID)
__addonpath__ = xbmc.translatePath(__addon__.getAddonInfo('path')).decode("utf-8")  # Translate path to change special:// protocol to a normal path
__addonicon__ = os.path.join(__addonpath__, 'icon.png')
__language__ = __addon__.getLocalizedString
__profile__ = xbmc.translatePath(__addon__.getAddonInfo('profile')).decode("utf-8")  # Translate path to change special:// protocol to a normal path


# Just a wrapper to keep icons and title easier consistent
def notify(text, duration=3000, title=None):
    if title == None:
        title = __language__(30000)
    
    xbmc.executebuiltin('Notification("{1}","{2}", {3}, {0})'.format(__addonicon__, title, text, duration))


def logDebug(msg):
    xbmc.log(msg, level=xbmc.LOGDEBUG)
    
def logError(msg):
    xbmc.log(msg, level=xbmc.LOGERROR)
    
def getConfiguredLampsList():
    lamps = []
    for i in range(1, hue.MAX_LAMPS+1):
        strId = str(i)

        if __addon__.getSetting("lamp" + strId) == "true":
            lamps.append(i)

    return lamps
 
# --- Settings related stuff ---
    
# Make sure profile path exists (it seems to be created when first time the settings are saved from the settings dialog)
try:
    os.makedirs(__profile__)
except OSError as exception:
    if exception.errno != errno.EEXIST:
        raise


    
class HueControlSettings:
    '''Settings dict stored in a file in the addon userdata folder'''
    
    def __init__(self):
        self.datafile = os.path.join(__profile__, huecontrol.SETTINGSFILE) # updated to 2 because of somehow the settings are incompatible and not sure how I can fix it
        self.data = {}

        if not os.path.isfile(self.datafile):
            self._createorupdatedefaultpresets()
        else:
            with open(self.datafile, 'rb') as handle:
                self.data = pickle.load(handle)
                
            if not 'settingsversion' in self.data:
                self._createorupdatedefaultpresets()
                self.data['settingsversion'] = 1
            
            if self.data['settingsversion'] == 1:
                if 'bridgeid' in self.data:
                    oldbridgeid = self.data['bridgeid']
                    self.data['bridgeid'] = "{0}fffe{1}".format(oldbridgeid[0:6], oldbridgeid[6:12]).lower()
                self._createorupdatedefaultpresets()
                self.data['settingsversion'] = 2

    def store(self):
        with open(self.datafile, 'wb') as handle:
            pickle.dump(self.data, handle)
            return True
        
        return False
    
    
    def _createorupdatedefaultpreset(self, presetName, state):
        if not presetName in self.data:
            self.data[presetName] = {'lights': {}}
        elif isinstance(self.data[presetName], basestring):
            if self.data[presetName] != "":
                self.data[presetName] = json.loads(self.data[presetName])
            else:
                self.data[presetName] = {}

        for i in range(1, hue.MAX_LAMPS+1):
            lightId = str(i)
            
            if not lightId in self.data[presetName]['lights']:
                self.data[presetName]['lights'][lightId] =  state

 
    def _createorupdatedefaultpresets(self):
        
        self._createorupdatedefaultpreset("scenePlaying", {'state':{'on':False}})
        self._createorupdatedefaultpreset("scenePaused",  {'state':{'on':True, 'bri':100}})
        self._createorupdatedefaultpreset("scenePreset1", {'state':{'on':True, 'bri':50,  'colormode':'xy', "xy":[0.5268,0.4133]}})
        self._createorupdatedefaultpreset("scenePreset2", {'state':{'on':True, 'bri':100, 'colormode':'xy', "xy":[0.4883,0.4149]}})
        self._createorupdatedefaultpreset("scenePreset3", {'state':{'on':True, 'bri':150, 'colormode':'xy', "xy":[0.4412,0.4055]}})
        self._createorupdatedefaultpreset("scenePreset4", {'state':{'on':True, 'bri':200, 'colormode':'xy', "xy":[0.3876,0.3811]}})
        self._createorupdatedefaultpreset("scenePreset5", {'state':{'on':True, 'bri':250, 'colormode':'xy', "xy":[0.3312,0.3400]}})

        self.store()
            
            
            

hue.py
#
# Code to control the hue bridge
#

import json
import httplib
import traceback
from datetime import datetime
from xml.dom import minidom
from threading import Lock
from threading import Thread
from Queue import Queue
import socket
import time
import re
import sys
import requests

MAX_LAMPS = 63   # Max lamps supported by the bridge
NUM_THREADS = 20 # Number of threads spawned when searching the network for bridges

class BridgeLocator:
    '''Class that finds hue bridges in the network'''
    bridges = []
    lock = Lock()
    q = Queue()

    def __init__(self, iprange=None, logfunc=None):
        self.iprange = iprange
        self.logfunc = logfunc
        
        self.log("BridgeLocator iprange:{0}".format(iprange))

    def FindBridgeTask(self):
        done = False
        while not done:
            item = self.q.get()
            self.log("Check IP: {0}".format(item))

            sys.stdout.write('.')
            if item == "STOP":
                done = True
            else:
                bridge = self.GetBridgeFromIp(item)
                if bridge:
                    with self.lock:
                        #self.bridges.append(bridge)
                        self.bridgesById[bridge.id] = bridge;

            self.q.task_done()

    def SearchIpRange(self, iprange=None):
    
        if iprange == None:
            iprange = self.iprange;

        self.log("SearchIpRange {0}".format(iprange))

        tmp = iprange.rfind('.')
        ipstart = iprange[0:tmp]
        for i in range(1,254):
            self.log("Adding IP address {0}.{1}".format(ipstart,i))
            self.q.put("{0}.{1}".format(ipstart, i))
        
    def FindBridges(self, progress=None):
        '''Crude first implementation, just try all addresses in the subnet'''
        #self.bridges = []
        self.bridgesById = {}

        # First try nupnp
        r = requests.get('http://www.meethue.com/api/nupnp', timeout=2)
        if r.status_code == 200:
            bridgelist = r.json()
            
            for nupnp_bridge in bridgelist:
                self.log(str(nupnp_bridge))
                
                # Ignore the ID, that will be fetched from the bridge if it exists at the given IP address
                bridge = self.GetBridgeFromIp(nupnp_bridge["internalipaddress"])
                if bridge:
                    #self.bridges.append(bridge);
                    self.bridgesById[bridge.id] = bridge;
        

        # Now start scan on local network
        for i in range(NUM_THREADS):
            t = Thread(target=self.FindBridgeTask)
            t.daemon = True
            t.start()

        rangecount = 0

        if (self.iprange is None):
            for ipaddress in [ip for ip in socket.gethostbyname_ex(socket.gethostname())[2] if not ip.startswith("127.0.0.")]: # magic from the internet (works on my windows, not on OpenElec)
            
                # Only spam valid home addres ranges
                
                # 10.0.0.0 through 10.255.255.255
                # 169.254.0.0 through 169.254.255.255 (Autoip)
                # 172.16.0.0 through 172.31.255.255
                # 192.168.0.0 through 192.168.255.255 
                if (not re.match("^127\.\d{123}\.\d{123}\.\d{123}$", ip) and 
                    not re.match("^10\.\d{123}\.\d{123}\.\d{123}$", ip) and 
                    not re.match("^192\.168.\d{123}$", ip) and 
                    not re.match("^172\.(1[6-9]|2[0-9]|3[0-1])\.[0-9]{123}\.[0-9]{123}$", ip) ):
                    self.SearchIpRange(ipaddress)
                    rangecount += 1
        else:
            self.SearchIpRange()
            rangecount = 1
        
        if (not progress is None):
            numitemstotal = rangecount * 254

            while (not self.q.empty()):
                numitemsleft = self.q.qsize()
                percent = ((numitemstotal - numitemsleft) / float(numitemstotal)) * 100
                progress(int(percent))
                time.sleep(0.3)
        
            progress(100)    # Done ;-)

        self.q.join()
        
        # Done with the threads, time to close them down, otherwise they hang Kodi when tying to close it
        for i in range(NUM_THREADS):
            self.q.put("STOP")

        # Convert to a list
        bridges = []
        for bridgeid in self.bridgesById.keys():
            bridges.append(self.bridgesById[bridgeid]);
        
        return bridges    


    def GetBridgeFromIp(self, ip):
        self.log("GetBridgeFromIp {0}".format(ip))

        bridge = None
        try:
            conn = httplib.HTTPConnection(ip, timeout=1)
            conn.request("GET", '/api/nouser/config') 
            resp = conn.getresponse()
            data = resp.read()

            self.log("{0}, {1}, {2}".format(ip, resp.status, data))
            
            if (resp.status == 200):
                # Got a config back. Parse it to get the bridge ID (and name)
                config = json.loads(data)
                
                # Older versions of the addon used the "serial" found in the description.xml this seems to be the macaddress without the colons.
                # The new software with API 1.10 and up have a "bridgeid" in /config, this seems to be the macaddress with "fffe" stuffed in the middle, use that if available (and use the lowercase as on the /nupnp)
                # If not available use the "mac" attribute in config to be able to work with older bridge software versions
                bridgeid = None
                if "bridgeid" in config:
                    bridgeid = config["bridgeid"].lower()
                elif "mac" in config:
                    self.log("Found a bridge with very old software version({0}) at IP {1}. You might want to update the bridge software with the Philips Hue app.".format(config["swversion"], ip))
                    mac = config["mac"].translate(None, " :")
                    bridgeid = "{0}fffe{1}".format(mac[0:6], mac[6:12]).lower()
                else:
                    self.log("Unable to detect bridge ID! Try updating the bridge software with the Philips Hue app")
                
                # An old (round) bridge can have been migrated to a new (square) bridge.
                # store the replacesbridgeid to ba able to detect migrated bridges for seamless transition
                replacesbridgeid = None
                if "replacesbridgeid" in config and config["replacesbridgeid"] != None:
                    replacesbridgeid = config["replacesbridgeid"].lower()

                friendlyname = config["name"]
                
                bridge = Bridge(ip, bridgeid, friendlyname, replacesbridgeid=replacesbridgeid)

            conn.close()
            
        except Exception as e:
            self.log("{0} Could not connect to {1}\n{2}".format(str(datetime.now()), ip, e))
            #self.log(traceback.print_exc())
            pass
    
        return bridge
                
        
    def FindBridgeById(self, bridgeid, lastip=None):
       
        # Try last known IP if provided
        if lastip:
            bridge = self.GetBridgeFromIp(lastip)
            if bridge and bridge.id == bridgeid:
                return bridge;
                
        # Not found at last IP just find all and filter.
        bridges = self.FindBridges()

        # First try to find an exact match
        for bridge in bridges:
            if bridge.id == bridgeid:
                return bridge;

        # See if there is a bridge that was migrated from the desired ID
        for bridge in bridges:
            if bridge.replacesbridgeid == bridgeid:
                # TODO update the stored bridge ID?
                return bridge;
        
        return None

    def log(self, msg):
        if self.logfunc:
            self.logfunc(msg)
        

class Bridge:
    '''Represents the bridge, use it to control the lights'''
    ip = 0
    id = 0
    name = "Philips hue"
    username = None
    devicetype = None
    authorized = False
    

    def __init__(self, ip=None, id=None, name=None, devicetype=None, username=None, logfunc=None, replacesbridgeid=None):
        self.ip = ip
        self.id = id
        self.name = name
        self.devicetype = devicetype
        self.username = username
        self.logfunc = logfunc
        self.replacesbridgeid = replacesbridgeid

    def __repr__(self):
        return "{0} - {1} - {2}".format(self.name, self.id, self.ip)
        

    def authorize(self):
        # Attempt to create a user
        data = {'devicetype': self.devicetype}
        
        #jsonstr = json.dumps(data)
        reply = self.POST("", data, addUsername=False)
        
        if ('success' in reply[0]):
            self.username = reply[0]["success"]["username"]  # Will be random name
            return 0
        elif ('error' in reply[0]):
            return reply[0]['error']['type']
        else:
            # Something weird happend
            return 666
        
    def isAuthorized(self):
        # Just get the config and check if we can see the whitelist
        reply = self.GET("/config")
        
        if ('whitelist' in reply):
            return 1

        return 0

    def getFullState(self):
        # Get the full state of the bridge
        reply = self.GET("")

        return reply; # json.dumps(reply)  # return the string so the receiver can easily store it (relevant for saving settings in Kodi which must be strings)
        
    def setFullStateLights(self, state, lightList=None, briOnly=False):
    
        # Set the lights back to the given full state of the bridge
        if isinstance(state, basestring):
            # Dont apply empty states
            if state == "":
                return
            parsedjson = json.loads(state)
        else:
            parsedjson = state
        
        if not 'lights' in parsedjson:
            return
            
        lights = parsedjson['lights']

        for i in range(1, MAX_LAMPS+1):
            strId = str(i)
            
            if  not lightList or (i in lightList):

                if (strId in lights):
                    storedstate = lights[strId]['state']
                    self.log("setFullStateLights ID:{0}, State:{1}".format(strId, storedstate))

                    lampstate = {}

                    lampstate['on'] = storedstate['on']
                    
                    # When lamp going to be off don't send the rest
                    # It avoids weird color flash
                    if storedstate['on'] != False:
                        lampstate['bri'] = storedstate['bri']
                        
                        if not briOnly:
                            # Also restore color stuff if available (e.g. living whites do not have colormode)
                            if 'colormode' in storedstate:
                                if (storedstate['colormode'] == 'ct'):
                                    lampstate['ct'] = storedstate['ct']
                                elif (storedstate['colormode'] == 'xy'):
                                    lampstate['xy'] = storedstate['xy']
                                elif (storedstate['colormode'] == 'hs'):
                                    lampstate['hue'] = storedstate['hue']
                                    lampstate['sat'] = storedstate['sat']

                    self.PUT('/lights/{0}/state'.format(strId), lampstate)

        
    def setLightOn(self, id):
        self.PUT('/lights/{0}/state'.format(id), '{"on": true}')
      
    def setLightOff(self, id):
        self.PUT('/lights/{0}/state'.format(id), {"on": False})
      

      
    def CLIP(self, method, resource, body, addUsername=True):

        reply = ""

        if type(body) == dict:
            body = json.dumps(body)

        url = '/api'    
        if addUsername == True:
            url += '/{0}'.format(self.username)
        url += resource

        self.log('> {0}, {1}, {2}\n'.format(method, url, body))

        try:
            r = requests.request(method, "http://{0}{1}".format(self.ip, url), data=body, timeout=1.5)
            reply = r.json()
        except requests.exceptions.ConnectTimeout as e:
            self.log("Connection timeout! IP:{0}, URL:{1}".format(self.ip, url))
        except Exception as e:
            self.log('E {0}\n'.format(traceback.format_exc()))
            raise e
        else:
            self.log('< {0}\n'.format(reply))
        
        return reply
        

    def GET(self, resource):
        return self.CLIP("GET", resource, "")

        
    def PUT(self, resource, body):
        return self.CLIP("PUT", resource, body)

        
    def POST(self, resource, body, addUsername=True):
        return self.CLIP("POST", resource, body, addUsername)

        
    def DELETE(self, resource):
        return self.CLIP("DELETE", resource, "")
        
    def log(self, msg):
        if self.logfunc:
            self.logfunc(msg)
 

    